src/components/notes/saving-indicator.jsx

```diff
@@ -19,6 +20,61 @@ const outOfSyncAfter = 30 * sec
  * @property {Ref} bannerRef - Where to display banners
  */
 
+/**
+ * Checks if the server is out of sync
+ * (our changes were not successfully sent and accepted for too much time)
+ * @param {object} params
+ * @param {boolean} params.isDirty - if there are changes to be sent
+ * @param {integer} params.dirtyAge - number of milliseconds waiting
+ *                                    with changes to be sent
+ * @returns {boolean}
+ */
+function checksOutOfSync({ isDirty, dirtyAge }) {
+  return isDirty && dirtyAge && dirtyAge > outOfSyncAfter
+}
+
+/**
+ * Checks for disconnection (browser offline or out of sync)
+ *
+ * @param {object} params
+ * @param {boolean} params.isOutOfSync
+ * @returns {{isDisconnected, disconnectedAge}}
+ */
+function checksDisconnection({ isOutOfSync }) {
+  const now = new Date()
+  const isBrowserOffline = useBrowserOffline()
+  const [disconnectedSince, setDisconnectedSince] = useState()
+  const isDisconnected = !!disconnectedSince
+  useEffect(() => {
+    const disconnectedNow = isBrowserOffline || isOutOfSync
+    if (!!disconnectedNow !== !!isDisconnected) {
+      setDisconnectedSince(disconnectedNow ? new Date() : null)
+    }
+  }, [isBrowserOffline, isOutOfSync, setDisconnectedSince, isDisconnected])
+  const disconnectedAge = isDisconnected ? now - disconnectedSince : 0
+  return { isDisconnected, disconnectedAge }
+}
+
+/**
+ * Should we open the offline indicator tooltip?
+ *
+ * We open it for a few seconds when we go offline
+ * then close it (it is still accessible when hovering over)
```

Comment:
I find the usePeriodicRender used here, a bit cryptic. I think using an imperative API like Alert here may have been simpler.

Reply:
What would you expect ?

Reply:
For this kind of things, I am accustomed to doing things like:

```
if (propertyHasChanged) {
  showAlert('Something has changed', 5000)
}
```

- Here I find it a bit convoluted to use a usePeriodicRender to remember to re-render at the end of the display period for the alert to disappear. I find that it's low-level to care about `render` here.

- I find the imperative API more suited for fire and forget alerts